nomad/acl_endpoint.go

```diff
@@ -1987,3 +1987,324 @@ func (a *ACL) WhoAmI(args *structs.GenericRequest, reply *structs.ACLWhoAmIRespo
+		return fmt.Errorf("all servers should be running version %v or later to use ACL binding rules",
+			minACLBindingRuleVersion)
+	}
+
+	// Check management level permissions
+	if acl, err := a.srv.ResolveToken(args.AuthToken); err != nil {
+		return err
+	} else if acl == nil || !acl.IsManagement() {
+		return structs.ErrPermissionDenied
+	}
+
+	// Validate non-zero set of binding rules. This must be done outside the
+	// validate function as that uses a loop, which will be skipped if the
+	// length is zero.
+	if len(args.ACLBindingRules) == 0 {
+		return structs.NewErrRPCCoded(http.StatusBadRequest, "must specify as least one binding rule")
+	}
+
+	stateSnapshot, err := a.srv.State().Snapshot()
+	if err != nil {
+		return err
+	}
+
+	// Validate each binding rules and compute the hash.
+	for idx, bindingRule := range args.ACLBindingRules {
+
+		if err := bindingRule.Validate(); err != nil {
+			return structs.NewErrRPCCodedf(http.StatusBadRequest, "binding rule %d invalid: %v", idx, err)
+		}
+
+		// If the caller has passed a rule ID, this call is considered an
+		// update to an existing rule. We should therefore ensure it is found
+		// within state.
+		if bindingRule.ID != "" {
+			existingBindingRule, err := stateSnapshot.GetACLBindingRule(nil, bindingRule.ID)
+			if err != nil {
+				return structs.NewErrRPCCodedf(
+					http.StatusInternalServerError, "binding rule lookup failed: %v", err)
+			}
+			if existingBindingRule == nil {
+				return structs.NewErrRPCCodedf(
+					http.StatusBadRequest, "cannot find binding rule %s", bindingRule.ID)
+			}
+		}
+
+		// Ensure the auth method linked to exists within state.
+		method, err := stateSnapshot.GetACLAuthMethodByName(nil, bindingRule.AuthMethod)
+		if err != nil {
+			return err
+		}
+		if method == nil {
+			return structs.NewErrRPCCodedf(
+				http.StatusBadRequest, "ACL auth method %s not found", bindingRule.AuthMethod)
+		}
+
+		// All the validation has passed, we can now canonicalize the object
+		// with the final internal data and set the hash.
+		bindingRule.Canonicalize()
+		bindingRule.SetHash()
+	}
+
+	// Update via Raft.
+	out, index, err := a.srv.raftApply(structs.ACLBindingRulesUpsertRequestType, args)
+	if err != nil {
+		return err
+	}
+
+	// Check if the FSM response, which is an interface, contains an error.
+	if err, ok := out.(error); ok && err != nil {
+		return err
+	}
+
+	// Populate the response. We do a lookup against the state to pick up the
+	// proper create / modify indexes.
+	stateSnapshot, err = a.srv.State().Snapshot()
+	if err != nil {
+		return err
+	}
+	for _, bindingRule := range args.ACLBindingRules {
+		lookupBindingRule, err := stateSnapshot.GetACLBindingRule(nil, bindingRule.ID)
+		if err != nil {
+			return structs.NewErrRPCCodedf(400, "ACL binding rule lookup failed: %v", err)
+		}
+		if lookupBindingRule != nil {
+			reply.ACLBindingRules = append(reply.ACLBindingRules, lookupBindingRule)
+		}
```

Comment:
State store read methods don't usually return an error unless its internal. And in any case if we get either an error or a nil value for this query, that seems like an internal error (5xx) to me rather than a request error (4xx) -- we should have just written this value so if we didn't, something went very wrong!

Reply:
Agreed on both counts, thanks! Fixed within https://github.com/hashicorp/nomad/pull/15529/commits/4754d09f3d1f935e186d336a913f7ba40b4b78f3